File: app/domain/layout.py
from uuid import uuid4
from app.domain.button import Button


class Layout:
    uuid: str
    name: str
    rows: int
    columns: int
    buttons: list[Button]
    
    def __init__(self, name: str, rows: int, columns: int, buttons: list[Button], uuid=None):
        self.name = name
        self.rows = rows
        self.columns = columns
        self.buttons = buttons
        if uuid is None:
            self.uuid = str(uuid4())
        else:
            self.uuid = uuid
    
    def to_dict(self):
        return {
            "uuid": self.uuid,
            "name": self.name,
            "rows": self.rows,
            "columns": self.columns,
            "buttons": [button.to_dict() for button in self.buttons]
        }
        
    def from_dict(self, data):
        self.uuid = data["uuid"]
        self.name = data["name"]
        self.rows = data["rows"]
        self.columns = data["columns"]
        self.buttons = [Button.from_dict(button) for button in data["buttons"]]
        return self

File: kahldeck-dashboard/src/domain/Layout.ts
import { v4 as uuidv4 } from 'uuid';
import { Button } from './Button';

export class Layout {
    readonly uuid: string;
    name: string;
    rows: number;
    columns: number;
    buttons: Button[];

    constructor(name: string, rows: number, columns: number, buttons: Button[], uuid?: string) {
        this.name = name;
        this.rows = rows;
        this.columns = columns;
        this.buttons = buttons;
        this.uuid = uuid || uuidv4();
    }

    toJSON(): Record<string, any> {
        return {
            uuid: this.uuid,
            name: this.name,
            rows: this.rows,
            columns: this.columns,
            buttons: this.buttons.map(button => {
                return button
            }
            )
        };
    }

    static fromJSON(data: Record<string, any>): Layout {
        return new Layout(
            data.name,
            data.rows,
            data.columns,
            data.buttons.map((buttonData: any) => Button.fromJSON(buttonData)),
            data.uuid
        );
    }

    public toString = (): string => {
        return `Layout(name=${this.name}, rows=${this.rows}, columns=${this.columns}, buttons=${this.buttons})`;
    }
}